trap.go

```diff
@@ -169,19 +169,24 @@ func (t *TrapListener) Listening() <-chan bool {
 }
 
 // Close terminates the listening on TrapListener socket
-//
-// NOTE: the trap code is currently unreliable when working with snmpv3 - pull requests welcome
 func (t *TrapListener) Close() {
-	// Prevent concurrent calls to Close
 	if atomic.CompareAndSwapInt32(&t.finish, 0, 1) {
-		// TODO there's bugs here
+		t.Lock()
+		defer t.Unlock()
+
 		if t.conn == nil {
 			return
 		}
-		if t.conn.LocalAddr().Network() == udp {
-			t.conn.Close()
+
+		if err := t.conn.Close(); err != nil {
+			t.Params.Logger.Printf("failed to Close() the TrapListener socket: %s", err)
+		}
+
+		select {
+		case <-t.done:
+		case <-time.After(time.Second * 3): // A timeout can prevent blocking forever
```

Comment:
Should we have this configurable in `TrapListener`?

Reply:
Good idea. Would adding a property named `CloseTimeout` of `time.Duriation`, with a default of 3, make sense for this?

Reply:
Thank you Ben, added the changes as discussed.